### PR TITLE
Merkle Proving

### PR DESCRIPTION
**Pull request by tchataigner**
## Description

This Pr introduces a gadget to prove a leaf inclusion in a Merkle Tree.

## Current progress

Used [Jellyfish Merkle Tree paper](https://developers.diem.com/papers/jellyfish-merkle-tree/2021-01-14.pdf) to implement a Proof of Inclusion algorithm.

It has a generic approach to which hashing algorithm is used, and currently natively handles the Keccak and Sha3 implementation in `bellpepper-keccak`.

## Links to any relevant issues or pull requests

Related to [linear#29](https://linear.app/lurk-lab/issue/LUR-29/proof-of-inclusion-gadget)

### CODE REVIEW

File: crates/merkle-inclusion/LICENSE-APACHE [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+      To apply the Apache License to your work, attach the following
+      boilerplate notice, with the fields enclosed by brackets "[]"
+      replaced with your own identifying information. (Don't include
+      the brackets!)  The text should be enclosed in the appropriate
+      comment syntax for the file format. We also recommend that a
+      file or class name and description of purpose be included on the
+      same "printed page" as the copyright notice for easier
+      identification within third-party archives.
+
+   Copyright 2023 Saravanan Vijayakumaran
```
Comment (avras) [inline crates/merkle-inclusion/LICENSE-APACHE:189]: You may want to remove my name "Saravanan Vijayakumaran" from this line.

File: crates/merkle-inclusion/LICENSE-MIT [hunk 2]
```diff
@@ -0,0 +1,21 @@
+MIT License
+
+Copyright (c) 2023 Saravanan Vijayakumaran
```
Comment (avras) [inline crates/merkle-inclusion/LICENSE-MIT:3]: Again my name here

File: crates/merkle-inclusion/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+fn hash_combine<E, CS>(
+    mut cs: CS,
+    hash1: HashValue,
+    hash2: HashValue,
+) -> Result<HashValue, SynthesisError>
+where
+    E: PrimeField,
+    CS: ConstraintSystem<E>,
+{
+    sha3(cs.namespace(|| "hash combine"), &[hash1, hash2].concat())
```
Comment (avras) [inline crates/merkle-inclusion/src/lib.rs:152]: I may be missing context for this pull request. Can the `sha3` here be replaced with a generic hash function circuit? For example, someone may want to use Poseidon to construct a Merkle tree and prove inclusion.

File: crates/merkle-inclusion/LICENSE-APACHE (hunk 1 above)
Comment (tchataigner) [reply crates/merkle-inclusion/LICENSE-APACHE:189]: Yes sorry, this is still a draft evolving :)

File: crates/merkle-inclusion/src/lib.rs (hunk 3 above)
Comment (tchataigner) [reply crates/merkle-inclusion/src/lib.rs:156]: For sure, I'm currently working on a commit that's dedicated to it. Should be pushed quite soon :+1:

**Comment by tchataigner** [comment]
@huitseeker , I added a test over the number of constraints in 
[437f1c8](https://github.com/lurk-lab/bellpepper-gadgets/pull/16/commits/437f1c89ba71b9af910e41ffa82932cba33679b7) as you suggested. I think I got it right :thinking: 

For a leaf at depth 3 of a tree it represents 455810 constraints for 256 bits long digests:
- 1283 for our inputs (targeted leaf key & hash, siblings, expected root)
- 454527 produced through the hashes happening in circuit and the expected vs actual root equality.

File: crates/merkle-inclusion/Cargo.toml [hunk 4]
```diff
@@ -0,0 +1,23 @@
+[package]
+name = "bellpepper-merkle-inclusion"
+version = "0.1.0"
+edition = "2021"
+homepage.workspace = true
+repository.workspace = true
```
Comment (huitseeker) [inline crates/merkle-inclusion/Cargo.toml:6]: Please fill in author, license, description.

File: crates/merkle-inclusion/src/traits.rs [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+/// This trait defines the necessary methods for computing hash digests within constraint systems.
+pub trait GadgetDigest<E: PrimeField> {
+    /// Output size for our hasher, in bits.
+    const OUTPUT_SIZE: usize;
+
+    type OutOfCircuitHasher: digest::Digest;
+
+    /// Get output size of the hasher.
+    fn output_size() -> usize {
+        Self::OUTPUT_SIZE
```
Comment (huitseeker) [inline crates/merkle-inclusion/src/traits.rs:18]: It would probably be useful to `debug_assert!` that this value is equal to `<Self::OutOfCircuitHasher as digest::Digest>::output_size()`.

File: crates/merkle-inclusion/tests/gadget.rs [hunk 6]
```diff
@@ -0,0 +1,370 @@
...
+use bellpepper_core::test_cs::TestConstraintSystem;
+use bellpepper_merkle_inclusion::traits::{GadgetDigest, Keccak, Sha3};
+use bellpepper_merkle_inclusion::{verify_proof, Leaf, Proof};
+use bitvec::order::Lsb0;
+use bitvec::prelude::BitVec;
+use bls12_381::{Bls12, Scalar};
+use digest::{Digest, Output};
+use pairing::Engine;
+
+macro_rules! create_hash_tests {
```
Comment (huitseeker) [inline crates/merkle-inclusion/tests/gadget.rs:13]: Those tests are under-using the trait you're defining. Arguably, once you have the trait defined, macros are overkill for what you want to do with tests here.

You could, for example, do the following:
 ```rust
 fn verify_inclusion_merkle<H: GadgetDigest<Scalar>>() {
     ...
     construct_merkle_tree::<H::OutOfCirCuitHasher>();
     ...
}

#[test]
fn test_verify_inclusion_merkle() {
    verify_inclusion_merkle::<Sha3>();
    verify_inclusion_merkle::<Keccak>();
    ...
}
```

File: crates/merkle-inclusion/src/traits.rs [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+            ) -> Result<Vec<Boolean>, SynthesisError> {
+                $digest_method(cs, input)
+            }
+        }
+    };
+}
+
+// Example use of the macro with OutOfCircuitHasher specified
+create_gadget_digest_impl!(Sha3, sha3, 256, Sha3_256);
+create_gadget_digest_impl!(Keccak, keccak256, 256, Keccak256);
```
Comment (huitseeker) [inline crates/merkle-inclusion/src/traits.rs:57]: it would be great to check we're not hard-coding the length to be 256, and an easy way to do this is to test with the neighboring SHA512 crate (which output is 512 bits)

File: crates/merkle-inclusion/Cargo.toml [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+
+[dependencies]
+ff = { workspace = true }
+bellpepper-core = { workspace = true }
+bellpepper = { workspace = true }
+bellpepper-keccak = { path="../keccak", version = "0.1.0" }
+hybrid-array = "0.2.0-rc.0"
+digest = "0.11.0-pre.4"
+sha3 = "0.11.0-pre.1"
```
Comment (huitseeker) [inline crates/merkle-inclusion/Cargo.toml:17]: should this be a dev-dependency?

File: crates/merkle-inclusion/Cargo.toml [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+homepage.workspace = true
+repository.workspace = true
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+
+[dependencies]
+ff = { workspace = true }
+bellpepper-core = { workspace = true }
+bellpepper = { workspace = true }
+bellpepper-keccak = { path="../keccak", version = "0.1.0" }
```
Comment (huitseeker) [inline crates/merkle-inclusion/Cargo.toml:14]: should this be a dev-dependency?

File: crates/merkle-inclusion/Cargo.toml [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+hybrid-array = "0.2.0-rc.0"
```
Comment (huitseeker) [inline crates/merkle-inclusion/Cargo.toml:15]: Where is this used?

File: crates/merkle-inclusion/tests/gadget.rs [hunk 11]
```diff
@@ -0,0 +1,370 @@
...
+/// - The root node is at index 14.
+///
+/// # Example
+/// ```
+/// let (root_hash, leaf_hashes_and_keys) = construct_merkle_tree();
+/// for (hash, key) in leaf_hashes_and_keys.iter() {
+///     // Process each leaf and intermediary node's hash and key...
+/// }
+/// ```
+pub fn construct_merkle_tree<D: Digest>() -> SimpleMerkleTree {
```
Comment (huitseeker) [inline crates/merkle-inclusion/tests/gadget.rs:315]: It's not needed for this iteration, but we'll probably want to open an issue to use proptest to generate this sort of test instance.

File: crates/merkle-inclusion/src/traits.rs (hunk 7 above)
Comment (huitseeker) [reply crates/merkle-inclusion/src/traits.rs:57]: This would also help with endianness: SHA512 uses big-endian, whereas the keccak crate is little-endian.

File: crates/merkle-inclusion/tests/gadget.rs (hunk 6 above)
Comment (tchataigner) [reply crates/merkle-inclusion/tests/gadget.rs:13]: Oh yes, completely missed that. Way cleaner!

File: crates/merkle-inclusion/tests/gadget.rs (hunk 11 above)
Comment (tchataigner) [reply crates/merkle-inclusion/tests/gadget.rs:315]: @huitseeker I just added it to have a clean tree to use for testing. Would you rather have me removing this implementation and implement a more straight forward construction of the leafs / proofs?

File: crates/merkle-inclusion/Cargo.toml (hunk 9 above)
Comment (tchataigner) [reply crates/merkle-inclusion/Cargo.toml:14]: So far I have implemented thje trait `GadgetDigest` for both `Keccak` and `Sha3` in this crate directly. Which mean I depend on the `bellpepper-keccak` crate to access the hashing algorithm entry point. 

This could be changed if I was to directly implement the trait in `bellpepper-keccak` crate, but I understood that this is not what we wanted.

File: crates/merkle-inclusion/tests/gadget.rs (hunk 11 above)
Comment (huitseeker) [reply crates/merkle-inclusion/tests/gadget.rs:315]: No, it's ok. I'm just mentioning specifically that `proptest_recurse` can generate random tree-structured data:
https://docs.rs/proptest-recurse/latest/proptest_recurse/
This is probably something we'll want, but in the future.

File: crates/merkle-inclusion/src/lib.rs [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+        HASH_LENGTH,
+        proof.siblings.len(),
+    );
+
+    // Reconstruct the root hash from the leaf and sibling hashes
+    let mut actual_root_hash = proof.leaf().hash().to_vec();
+
+    for (i, (sibling_hash, bit)) in proof
+        .siblings()
+        .to_vec()
```
Comment (huitseeker) [inline crates/merkle-inclusion/src/lib.rs:95]: Why do you need to `to_vec()` before iter? Would `into_iter()` not suffice?

File: crates/merkle-inclusion/src/lib.rs [hunk 13]
```diff
@@ -0,0 +1,158 @@
...
+    mut cs: CS,
+    expected_root: Vec<Boolean>,
+    proof: Proof,
+) -> Result<Vec<Boolean>, SynthesisError>
+where
+    E: PrimeField,
+    CS: ConstraintSystem<E>,
+    GD: GadgetDigest<E>,
+{
+    assert_eq!(expected_root.len(), GD::output_size());
```
Comment (huitseeker) [inline crates/merkle-inclusion/src/lib.rs:80]: Is there a missing check on the length of the key here?

File: crates/merkle-inclusion/src/traits.rs (hunk 7 above)
Comment (tchataigner) [reply crates/merkle-inclusion/src/traits.rs:57]: Thanks, this is pretty helpful!

File: crates/merkle-inclusion/src/lib.rs [hunk 14]
```diff
@@ -0,0 +1,158 @@
...
+pub type HashValue = Vec<Boolean>;
+pub type Key = Vec<Boolean>;
+const HASH_LENGTH: usize = 32 * 8;
+
+/// Represents a leaf node in a Jellyfish Merkle Tree.
+///
+/// Each leaf node contains a key-value pair, where the key is used to determine the position in the tree and the value
+/// is the data stored.
+pub struct Leaf {
+    key: Key,
```
Comment (huitseeker) [inline crates/merkle-inclusion/src/lib.rs:18]: What's the expected endianness of this key? Shall we document it?

File: crates/merkle-inclusion/src/lib.rs [hunk 15]
```diff
@@ -0,0 +1,158 @@
...
+    proof: Proof,
+) -> Result<Vec<Boolean>, SynthesisError>
+where
+    E: PrimeField,
+    CS: ConstraintSystem<E>,
+    GD: GadgetDigest<E>,
+{
+    assert_eq!(expected_root.len(), GD::output_size());
+
+    //Assert that we do not have more siblings than the length of our hash (otherwise cannot know which path to go)
```
Comment (huitseeker) [inline crates/merkle-inclusion/src/lib.rs:82]: This is materializing an undocumented assumption that we use an index of size `HASH_LENGTH` to thread the sibling hashes. 

We use the leaf's key to thread sibling hashes.

1. Is there an undocumented assumption on the length of the key here?
2. what if `proof.leaf().key().len() > HASH_LENGTH`? what breaks in the algorithm.

File: crates/merkle-inclusion/src/lib.rs (hunk 14 above)
Comment (tchataigner) [reply crates/merkle-inclusion/src/lib.rs:18]: As of now I handle the key to be Little Endian all the time. I think it would actually make more sense if it could be of the same endian of the hashing algorithm it is being associated to. Will update the code to reflect it.

File: crates/merkle-inclusion/src/lib.rs (hunk 15 above)
Comment (tchataigner) [reply crates/merkle-inclusion/src/lib.rs:82]: So as you have seen, [we expect the key to be smaller or of same size as the output size of our hashing algorithm](https://github.com/lurk-lab/bellpepper-gadgets/pull/16/files#diff-260ab9bebaec6c9babe2300c428142d465ac60b357284985399dd94799bac24aR103). But you are right that we are missing a check here. I will add an assert and also add more tests to ensure that those edge cases are covered.

Comment (huitseeker) [reply crates/merkle-inclusion/src/lib.rs:82]: > we expect the key to be smaller or of same size as the output size of our hashing algorithm

Right, that's correct, the code expects the key to be at most as long as our hash output is.

The question I'm asking is: do we need to enshrine *that* assumption in the algorithm, and if so, why?

One *related* assumption that I am sure we need to make is that the key needs to be as long as the siblings size. The reason we need to make that assumption is that otherwise we can't index into the siblings array using the key.

The two assumptions are related, and the later assumption is clearly needed. Is the earlier assumption needed as well?

Comment (tchataigner) [reply crates/merkle-inclusion/src/lib.rs:82]: I'm sorry for the answer, I think I wrote it down without thinking it through at first.

To properly navigate our proof and ensure that it corresponds to a valid leaf we need to only need to ensure that `Len(proof.siblings) <= Len(proof.leaf.key)`. It verifies that we will get information on position for all of our siblings.

So you are right to say that the first assumption is not needed and for the second one I would slightly change it to accepts longer keys as it would not break out algorithm anyway.

File: crates/merkle-inclusion/src/lib.rs (hunk 14 above)
Comment (tchataigner) [reply crates/merkle-inclusion/src/lib.rs:18]: Did this in 775a5fa

File: crates/merkle-inclusion/src/lib.rs (hunk 15 above)
Comment (tchataigner) [reply crates/merkle-inclusion/src/lib.rs]: Fixing the iteration process done in 775a5fa.

Assertion for number of siblings and length of the key done in 1280301.

File: crates/merkle-inclusion/tests/gadget.rs (hunk 6 above)
Comment (tchataigner) [reply crates/merkle-inclusion/tests/gadget.rs]: Did so in 6057334.

File: crates/merkle-inclusion/src/traits.rs (hunk 7 above)
Comment (tchataigner) [reply crates/merkle-inclusion/src/traits.rs]: Done in 04ae409.

File: crates/merkle-inclusion/src/lib.rs (hunk 14 above)
Comment (tchataigner) [reply crates/merkle-inclusion/src/lib.rs:18]: Behavior needed correction, updated in a74aa26

File: crates/merkle-inclusion/src/lib.rs [hunk 16]
```diff
@@ -0,0 +1,186 @@
...
+        proof.siblings.len(),
+        proof.leaf().key().len(),
+    );
+    //Assert that we do not have more siblings than the length of our hash (otherwise cannot know which path to go)
+
+    // Reconstruct the root hash from the leaf and sibling hashes
+    let mut actual_root_hash = proof.leaf().hash().to_vec();
+
+    let key_iterator =
+        conditional_reverse::<_, _, GD>(proof.leaf().key().iter().take(proof.siblings().len()));
```
Comment (huitseeker) [inline crates/merkle-inclusion/src/lib.rs:95]: if I have for example :
- `proof.siblings.len() == 5` and 
- `proof.leaf().key().len() == 8`, and 
- `GD::is_little_endian() == true`

The sequence of indexes of `proof.leaf.key()` that will be used to thread `proof().siblings()` is `[4, 3, 2, 1, 0]`. Is that what the user should expect, or should they expect `[7, 6, 5, 4, 3]` instead? Is the MSB of the key not at index 7 in LE order?

I think there's two ways to go about this:
1. explicitly state everywhere that key and gadget should have the same endianness, and in that case, reverse first, and  `.take(proof.siblings.len())` after. This is closer to the Aptos spec (note the mention of MSB of the key there), but may be less simple for other Merkle algorithms.
2. not explicitly make any relation between the gadget's endianness and the key's iteration order, and say that in a proof, key and siblings should be iterable in *the same* order, whatever it is: to the point that the algorithm should operate on the iterator formed by `proof.leaf.key().iter().zip(proof.siblings.iter())` (after checking `proof.keaf.key().len() > proof.siblings().len()`)

(I don't particularly care which of these two solutions we use, but I am mindful that we put the user in a situation to know precisely which approach we're enforcing.)

File: crates/merkle-inclusion/src/lib.rs [hunk 17]
```diff
@@ -0,0 +1,186 @@
...
+/// Returns a boxed double-ended iterator over references to [`Boolean`]. The order of elements in the iterator will be
+/// reversed if [`GadgetDigest`] indicates little-endian order.
+fn conditional_reverse<
+    'a,
+    I: DoubleEndedIterator<Item = &'a Boolean> + 'a,
+    E: PrimeField,
+    GD: GadgetDigest<E>,
+>(
+    iter: I,
+) -> Box<dyn DoubleEndedIterator<Item = &'a Boolean> + 'a> {
```
Comment (huitseeker) [inline crates/merkle-inclusion/src/lib.rs:180]: It would be great if we could inline this and avoid the performance cost of `Box<dyn MyType>` for these few lines.

File: crates/merkle-inclusion/src/lib.rs [hunk 18]
```diff
@@ -0,0 +1,186 @@
...
+                    cs.namespace(|| format!("sibling {}", i)),
+                    &[actual_root_hash, sibling_hash.to_owned()].concat(),
+                )?
+            }
+        } else {
+            return Err(SynthesisError::Unsatisfiable);
+        }
+    }
+
+    hash_equality(cs, expected_root, actual_root_hash)
```
Comment (huitseeker) [inline crates/merkle-inclusion/src/lib.rs:115]: Aka, I believe:
```rust
for (i, (sibling_hash, bit)) in proof.siblings().iter().zip(key_iterator).enumerate() {
    let b = bit.get_value().ok_or(SynthesisError::Unsatisfiable)?;

    // Determine the order of hashing based on the bit value
    let hash_order = if b {
        vec![sibling_hash.to_owned(), actual_root_hash]
    } else {
        vec![actual_root_hash, sibling_hash.to_owned()]
    };

    // Compute the new hash
    actual_root_hash = GD::digest(cs.namespace(|| format!("sibling {}", i)), &hash_order.concat())?;
}

hash_equality(cs, expected_root, actual_root_hash)
```

**Review by huitseeker — COMMENTED** [review]
This is getting close, I think we need to improve how we document the assumptions we'd have on the key-siblings joint iteration. Thanks!

File: crates/merkle-inclusion/src/lib.rs (hunk 16 above)
Comment (tchataigner) [reply crates/merkle-inclusion/src/lib.rs:95]: I've went along with solution 2. , I think it should suffice in our context.

File: LICENSE-APACHE
Comment (huitseeker) [inline LICENSE-APACHE:1]: I think we don't need the License files replicated at the root of the repo if we copy them in each crate, which we have done so far.

I'd love to address those copies, but I'd want to do it in another PR, to keep this one focused on Merkle trees => I think this should not add any license files other than the ones in its own crate.

Comment (tchataigner) [reply LICENSE-APACHE:1]: alright, I'll remove the license files at the root.
